Validate content extensions whenever loading them in

pascoej opened the pull request:
#### 040e877236f583892af4d7d8f9d84318dd1ca7ee
<pre>
Validate content extensions whenever loading them in
<a href="https://bugs.webkit.org/show_bug.cgi?id=289293">https://bugs.webkit.org/show_bug.cgi?id=289293</a>
<a href="https://rdar.apple.com/145703008">rdar://145703008</a>

Reviewed by Alex Christensen.

We should perform some validation when adding extensions instead of
killing the web process if the file got corrupted for whatever reason.

Added API test.

* Source/WebCore/contentextensions/DFABytecodeInterpreter.h:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::openAndMapContentRuleList):
(API::ContentRuleListStore::corruptContentRuleListURLFilter):
* Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListStore.mm:
(-[WKContentRuleListStore _corruptContentRuleListURLFilterForIdentifier:]):
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListStorePrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKContentExtensionStore.mm:
(TEST_F(WKContentRuleListStoreTest, CorruptURLFilter)):

Canonical link: <a href="https://commits.webkit.org/291937@main">https://commits.webkit.org/291937@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4646032479cac5458f6997839a0ff628081dd957

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94451 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14043 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3824 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99470 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44978 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96501 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14339 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22471 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72081 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29404 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97453 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10661 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85280 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52411 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10355 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2972 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44293 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/3073 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101514 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21505 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15690 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/81082 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21754 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81300 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80457 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20054 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25003 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14731 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21482 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/26641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21166 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24626 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22907 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 reviewed (changes requested):
I don't think we should do this.

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -234,6 +236,19 @@ static std::optional<MappedData> openAndMapContentRuleList(const WTF::String& pa
+    auto universalActions = copyToVector(interpreter.actionsMatchingEverything());
+    for (uint64_t universalActionLocation : universalActions) {
+        if (universalActionLocation >= metaData->actionsSize) {
+            LOG(ContentRuleLists, "Universal action has location outside range of serialized actions. The compiled extension may be corrupted: %s", decodeFromFilename(path).utf8().data());
```
I don't think the decodeFromFilename is useful here.

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -234,6 +236,19 @@ static std::optional<MappedData> openAndMapContentRuleList(const WTF::String& pa
     auto metaData = decodeContentRuleListMetaData(fileData);
     if (!metaData)
         return std::nullopt;
+    const size_t headerAndSourceSize = headerSize(metaData->version) + metaData->sourceSize;
```
nit: I'd add a newline before this.  Actually, I'd put all this into a helper function.

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -660,6 +675,38 @@ void ContentRuleListStore::corruptContentRuleListHeader(const WTF::String& ident
+    size_t headerAndSourceSize = headerSize(metaData->version) + metaData->sourceSize;
+    size_t actionsOffset = headerAndSourceSize;
+    size_t urlFiltersOffset = actionsOffset + metaData->actionsSize;
+    size_t urlFilterLocationOffset = urlFiltersOffset + 5;
```
Could you give 5 a name or calculate it from other values so it's not a magic number?

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -660,6 +675,38 @@ void ContentRuleListStore::corruptContentRuleListHeader(const WTF::String& ident
     closeFile(file);
 }
 
+void ContentRuleListStore::corruptContentRuleListURLFilter(const WTF::String& identifier)
```
corruptContentRuleListActionsMatchingEverything?

pascoej committed:
Validate content extensions whenever loading them in
https://bugs.webkit.org/show_bug.cgi?id=289293
rdar://145703008

Reviewed by Alex Christensen.

We should perform some validation when adding extensions instead of
killing the web process if the file got corrupted for whatever reason.

Added API test.

* Source/WebCore/contentextensions/DFABytecodeInterpreter.h:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::openAndMapContentRuleList):
(API::ContentRuleListStore::corruptContentRuleListURLFilter):
* Source/WebKit/UIProcess/API/APIContentRuleListStore.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListStore.mm:
(-[WKContentRuleListStore _corruptContentRuleListURLFilterForIdentifier:]):
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListStorePrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKContentExtensionStore.mm:
(TEST_F(WKContentRuleListStoreTest, CorruptURLFilter)):

Canonical link: https://commits.webkit.org/291937@main
